File: VB/WindowsApplication1/MyDataSourceWrapper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Collections

Namespace WindowsApplication1
	Public Class MyDataSourceWrapper
		Implements ITypedList, IList

		Public ReadOnly NestedList As IList
		Public ReadOnly Property NestedTypedList() As ITypedList
			Get
				Return CType(NestedList, ITypedList)
			End Get
		End Property
		Private descriptor As MyPropertyDescriptor


		Public Sub New(ByVal dataSource As ITypedList, ByVal unboundColumnFieldName As String, ByVal getValueHandler As GetValueEventHanlder, ByVal setValueHanlder As SetValueEventHanlder)
			Me.NestedList = CType(dataSource, IList)
			descriptor = New MyPropertyDescriptor(unboundColumnFieldName)
			AddHandler descriptor.GetUnboundValue, getValueHandler
			AddHandler descriptor.SetUnboundValue, setValueHanlder
		End Sub

		#Region "ITypedList Members"

		Private Function GetItemProperties(ByVal listAccessors() As PropertyDescriptor) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
			Dim propertiesList As PropertyDescriptorCollection = NestedTypedList.GetItemProperties(listAccessors)
			propertiesList.Add(descriptor)
			Return propertiesList
		End Function

		Private Function GetListName(ByVal listAccessors() As PropertyDescriptor) As String Implements ITypedList.GetListName
			Return NestedTypedList.GetListName(listAccessors)
		End Function

		#End Region

		#Region "IList Members"

		Private Function Add(ByVal value As Object) As Integer Implements IList.Add
		   Return NestedList.Add(value)
		End Function

		Private Sub Clear() Implements IList.Clear
			NestedList.Clear()
		End Sub

		Private Function Contains(ByVal value As Object) As Boolean Implements IList.Contains
			Return NestedList.Contains(value)
		End Function

		Private Function IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
			Return NestedList.IndexOf(value)
		End Function

		Private Sub Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
			NestedList.Insert(index, value)
		End Sub

		Private ReadOnly Property IsFixedSize() As Boolean Implements IList.IsFixedSize
			Get
				Return NestedList.IsFixedSize
			End Get
		End Property

		Private ReadOnly Property IsReadOnly() As Boolean Implements IList.IsReadOnly
			Get
				Return NestedList.IsReadOnly
			End Get
		End Property

		Private Sub Remove(ByVal value As Object) Implements IList.Remove
			NestedList.Remove(value)
		End Sub

		Private Sub RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
			NestedList.RemoveAt(index)
		End Sub

		Public Property IList_Item(ByVal index As Integer) As Object Implements IList.Item
			Get
				Return NestedList(index)
			End Get
			Set(ByVal value As Object)
				NestedList(index)= value
			End Set
		End Property

		#End Region

		#Region "ICollection Members"

		Private Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
			NestedList.CopyTo(array, index)
		End Sub

		Private ReadOnly Property Count() As Integer Implements ICollection.Count
			Get
				Return NestedList.Count
			End Get
		End Property

		Private ReadOnly Property IsSynchronized() As Boolean Implements ICollection.IsSynchronized
			Get
				Return NestedList.IsSynchronized
			End Get
		End Property

		Private ReadOnly Property SyncRoot() As Object Implements ICollection.SyncRoot
			Get
				Return NestedList.SyncRoot
			End Get
		End Property

		#End Region

		#Region "IEnumerable Members"

		Private Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
			Return NestedList.GetEnumerator()
		End Function

		#End Region
	End Class

	Public Class MyPropertyDescriptor
		Inherits PropertyDescriptor
		Protected Sub New(ByVal name As String, ByVal attrs() As Attribute)
			MyBase.New(name, attrs)

		End Sub
		Protected Sub New(ByVal descr As MemberDescriptor)
			MyBase.New(descr)

		End Sub
		Protected Sub New(ByVal descr As MemberDescriptor, ByVal attrs() As Attribute)
			MyBase.New(descr, attrs)

		End Sub

		Public Sub New(ByVal name As String)
			MyBase.New(name, Nothing)

		End Sub



		Private Event onGetValue As GetValueEventHanlder

		Public Custom Event GetUnboundValue As GetValueEventHanlder
			AddHandler(ByVal value As GetValueEventHanlder)
				AddHandler onGetValue, value
			End AddHandler
			RemoveHandler(ByVal value As GetValueEventHanlder)
				RemoveHandler onGetValue, value
			End RemoveHandler
			RaiseEvent(ByVal component As Object, <System.Runtime.InteropServices.Out()> ByRef result As Object)
			End RaiseEvent
		End Event





		Private Event onSetValue As SetValueEventHanlder

		Public Custom Event SetUnboundValue As SetValueEventHanlder
			AddHandler(ByVal value As SetValueEventHanlder)
				AddHandler onSetValue, value
			End AddHandler
			RemoveHandler(ByVal value As SetValueEventHanlder)
				RemoveHandler onSetValue, value
			End RemoveHandler
			RaiseEvent(ByVal component As Object, ByVal value As Object)
			End RaiseEvent
		End Event


		Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
			Return False
		End Function

		Public Overrides ReadOnly Property ComponentType() As Type
			Get
				Return GetType(Object)
			End Get
		End Property

		Public Overrides Function GetValue(ByVal component As Object) As Object
			Dim result As Object = Nothing
			RaiseEvent onGetValue(component, result)
			Return result
		End Function

		Public Overrides ReadOnly Property IsReadOnly() As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property PropertyType() As Type
			Get
				Return GetType(Object)
			End Get
		End Property

		Public Overrides Sub ResetValue(ByVal component As Object)

		End Sub

		Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
			RaiseEvent onSetValue(component, value)
		End Sub

		Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
			Return False
		End Function
	End Class
			Public Delegate Sub GetValueEventHanlder(ByVal component As Object, <System.Runtime.InteropServices.Out()> ByRef result As Object)
		 Public Delegate Sub SetValueEventHanlder(ByVal component As Object, ByVal value As Object)
End Namespace

File: CS/WindowsApplication1/MyDataSourceWrapper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Collections;

namespace WindowsApplication1
{
    public class MyDataSourceWrapper : ITypedList, IList
    {

        public readonly IList NestedList;
		public ITypedList NestedTypedList { get { return (ITypedList)NestedList; } }
        MyPropertyDescriptor descriptor;


        public MyDataSourceWrapper(ITypedList dataSource, string unboundColumnFieldName, GetValueEventHanlder getValueHandler, SetValueEventHanlder setValueHanlder)
        {
            this.NestedList = (IList)dataSource;
            descriptor = new MyPropertyDescriptor(unboundColumnFieldName);
            descriptor.GetUnboundValue += getValueHandler;
            descriptor.SetUnboundValue += setValueHanlder;
        }

        #region ITypedList Members

        PropertyDescriptorCollection ITypedList.GetItemProperties(PropertyDescriptor[] listAccessors)
        {
            PropertyDescriptorCollection propertiesList = NestedTypedList.GetItemProperties(listAccessors);
            propertiesList.Add(descriptor);
            return propertiesList;
        }

        string ITypedList.GetListName(PropertyDescriptor[] listAccessors)
        {
            return NestedTypedList.GetListName(listAccessors);
        }

        #endregion

        #region IList Members

        int IList.Add(object value)
        {
           return NestedList.Add(value);
        }

        void IList.Clear()
        {
            NestedList.Clear();
        }

        bool IList.Contains(object value)
        {
            return NestedList.Contains(value);
        }

        int IList.IndexOf(object value)
        {
            return NestedList.IndexOf(value);
        }

        void IList.Insert(int index, object value)
        {
            NestedList.Insert(index, value);
        }

        bool IList.IsFixedSize
        {
            get { return  NestedList.IsFixedSize; }
        }

        bool IList.IsReadOnly
        {
            get { return NestedList.IsReadOnly; }
        }

        void IList.Remove(object value)
        {
            NestedList.Remove(value);
        }

        void IList.RemoveAt(int index)
        {
            NestedList.RemoveAt(index);
        }

        object IList.this[int index]
        {
            get
            {
                return NestedList[index];
            }
            set
            {
                NestedList[index]= value;
            }
        }

        #endregion

        #region ICollection Members

        void ICollection.CopyTo(Array array, int index)
        {
            NestedList.CopyTo(array, index);
        }

        int ICollection.Count
        {
            get { return NestedList.Count; }
        }

        bool ICollection.IsSynchronized
        {
            get { return NestedList.IsSynchronized; }
        }

        object ICollection.SyncRoot
        {
            get { return NestedList.SyncRoot; }
        }

        #endregion

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator()
        {
            return NestedList.GetEnumerator();
        }

        #endregion
    }

    public class MyPropertyDescriptor: PropertyDescriptor
    {
        protected MyPropertyDescriptor(string name, Attribute[] attrs)
            : base(name, attrs)
        {

        }
        protected MyPropertyDescriptor(MemberDescriptor descr)
            : base(descr)
        {
            
        }
        protected MyPropertyDescriptor(MemberDescriptor descr, Attribute[] attrs)
            : base(descr, attrs)
        {
            
        }

        public MyPropertyDescriptor(string name)
            : base(name, null)
        {

        }



        private GetValueEventHanlder onGetValue;

        public event GetValueEventHanlder GetUnboundValue
        {
            add
            {
                onGetValue += value;
            }
            remove
            {
                onGetValue -= value;
            }
        }





        private SetValueEventHanlder onSetValue;

        public event SetValueEventHanlder SetUnboundValue
        {
            add
            {
                onSetValue += value;
            }
            remove
            {
                onSetValue -= value;
            }
        }
  

        public override bool CanResetValue(object component)
        {
            return false;
        }

        public override Type ComponentType
        {
            get { return typeof(object); }
        }

        public override object GetValue(object component)
        {
            object result = null;
            if (onGetValue != null)
            {
                onGetValue(component, out result);
            }
            return result;
        }

        public override bool IsReadOnly
        {
            get { return false; }
        }

        public override Type PropertyType
        {
            get { return typeof(object); }
        }

        public override void ResetValue(object component)
        {
            
        }

        public override void SetValue(object component, object value)
        {
            if (onSetValue != null)
                onSetValue(component, value);
        }

        public override bool ShouldSerializeValue(object component)
        {
            return false;
        }
    }
            public delegate void GetValueEventHanlder(object component, out object result);
         public delegate void SetValueEventHanlder(object component, object value);
}
